pkg/prometheus/prometheus.go

```diff
@@ -0,0 +1,182 @@
+	StringResult MetricType = "string"
+
+	prometheusSchema = "http"
+	// todo: the namespace can be different from runai
+	namespace = "runai"
+	promLabel = "prometheus-operator-prometheus"
+
+)
+
+func BuildPromethuseClient(c kubernetes.Interface) (*Client, error) {
+
+	ps := &Client {
+		client: c,
+	}
+	service, err := ps.GetPrometheusService()
+	if err != nil {
+		return nil, err
+	}
+	ps.service = *service
+	
+	return ps, nil
+}
+
+
+func (ps *Client) GetPrometheusService() (service *v1.Service, err error) {
+
+	list, err := ps.client.CoreV1().Services(namespace).List( metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("app=%s", promLabel),
+	})
+
+	if err != nil {
+		return 
+	} else if len(list.Items) > 0 {
+		service = &list.Items[0]
+	} else {
+		return nil, fmt.Errorf("Not found a server for prometheus")
+	}
+
+	return
+}
+
+func (ps *Client)  Query( query string) (data MetricData,  err error) {
+	var rst *Metric
+
+	req := ps.client.CoreV1().Services(ps.service.Namespace).ProxyGet(prometheusSchema, ps.service.Name , "9090", "api/v1/query", map[string]string{
+		"query": query,
+		"time": strconv.FormatInt(time.Now().Unix(), 10),
+	})
+
+	log.Debugf("Query prometheus for by %s in ns %s", query, ps.service.Namespace)
+	metric, err := req.DoRaw()
+	if err != nil {
+		log.Debugf("Query prometheus failed due to err %v", err)
+		log.Debugf("Query prometheus failed due to result %s", string(metric))
+		return
+	}
+	rst = &Metric{}
+	err = json.Unmarshal(metric, rst)
+	log.Debugf("Prometheus metric:%v", rst)
+	if err != nil {
+		err = fmt.Errorf("failed to unmarshall heapster response: %v", err)
+		return
+	}
+	if rst.Status != SuccessStatus {
+		err = fmt.Errorf("failed to query prometheus, status: %s", rst.Status)
+		return 
+	}
+	if len(rst.Data.Result) == 0 {
+		log.Debugf("The metric is not exist in prometheus for query  %s", query)
+	}
+	data = rst.Data
+	return
+ }
+
+
+ // MultipuleQueriesToItemsMap map multipule queries to items by given itemId
+func (ps *Client) MultipuleQueriesToItemsMap(q MultiQueries, itemID string) ( ItemsMap, error) {
+	queryResults := map[string]MetricData{}
+	rst := ItemsMap{}
+	// funcs := []func() error{}
+	var mux sync.Mutex
+
+	var err error;
+
+	for queryName, query := range q {
+		getFunc := func() error {
+			rst, err := ps.Query(query)
+			mux.Lock()
+			queryResults[queryName] = rst
+			// fmt.Print(queryName,"\n\n",rst, "\n\n")
+			mux.Unlock()
+			return err
+		}
+		getFunc();
+		// funcs = append(funcs, getFunc)
+	}
+	// err = util.Parallel(funcs...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// map the result to items by the given 'itemId' 
+	for queryName, qr := range queryResults {
+		// todo: check the metric type
+		for _, metricResult := range qr.Result {
+			// search the itemId in metric labels
+			key, ok := metricResult.Metric[itemID]
+			if !ok {
+				return nil, fmt.Errorf("[Prometheos] Not found an 'itemID' (%s) on the metric query: %s => %s",itemID, queryName, q[queryName] )
```

Comment:
This error will later be shown to the user.
the user has no idea what is `itemID`. let's rephrase it

Reply:
first, the error not will happen even though the Prometheus is not available (just if we have a problem with the metrics)
the error will be wrapped in a nice message that Prometheus service not found 
I assume that if this error happens we want the message to understand what went wrong
do you agree?

Reply:
The problem is with the word `itemID`.
You can write 
```suggestion
				return nil, fmt.Errorf("[Prometheos] Failed to find key: (%s) on the metric query: %s => %s",itemID, queryName, q[queryName] )
```